src/coreclr/jit/codegenxarch.cpp

```diff
@@ -3311,6 +3316,52 @@ void CodeGen::genCodeForInitBlkUnroll(GenTreeBlk* node)
 #endif
 }
 
+//------------------------------------------------------------------------
+// genCodeForInitBlkLoop - Generate code for an InitBlk using an inlined for-loop.
+//    It's needed for cases when size is too big to unroll and we're not allowed
+//    to use memset call due to atomicity requirements.
+//
+// Arguments:
+//    initBlkNode - the GT_STORE_BLK node
+//
+void CodeGen::genCodeForInitBlkLoop(GenTreeBlk* initBlkNode)
+{
+    GenTree* const dstNode  = initBlkNode->Addr();
+    GenTree* const zeroNode = initBlkNode->Data();
+
+    genConsumeReg(dstNode);
+    genConsumeReg(zeroNode);
+
+    const regNumber dstReg  = dstNode->GetRegNum();
+    const regNumber zeroReg = zeroNode->GetRegNum();
+
+    //  xor      zeroReg, zeroReg
+    //  mov      qword ptr [dstReg], zeroReg
+    //  mov      offsetReg, <block size>
+    //.LOOP:
+    //  mov      qword ptr [dstReg + offsetReg], zeroReg
+    //  sub      offsetReg, 8
+    //  jne      .LOOP
```

Comment:
Couldn't you get rid of the initial `mov qword ptr [dstReg], zeroReg` by using `jns` instead of `jne`?

e.g., https://godbolt.org/z/Y7bEGzrfx

Reply:
I think it serves as a nullcheck here, if we don't have it then for some large size we'll start our "reversed" zeroing from `null + largeConstant` and may not be able to handle that as NRE? (will be AVE instead)

Reply:
If it is a required null check, make sure there is a comment here that makes that clear. (And if there are any conditions under which the null check can be omitted?)

Reply:
Added a comment, we probably can detect cases where we don't need the nullcheck, but it also does the useful work - it clears the first pointer (the loop doesn't clear it again due to `JUMP_NE`) so it's probably not worth the effort to try to skip it since diffs aren't too big and, presumably, we only emit loops for cases when we don't unroll - large sizes, so most of the performance is spent inside the loop.